File: src/chapters/differential-equations/content/code/taylor2.cpp
#include <iostream>
#include <functional>
using namespace std;
using Fn = function<double(double, double)>;

// Second-order Taylor method; df is the total derivative f' = f_t + f_y f.
double taylor2(Fn f, Fn df, double t0, double y0, double h, int n) {
    double t = t0, y = y0;
    for (int i = 0; i < n; ++i) { y += h * f(t, y) + h * h / 2 * df(t, y); t += h; }
    return y;
}

int main() {
    auto f = [](double t, double y) { return 2 * y - 10 * t * t + 2 * t; };
    auto df = [](double t, double y) { return 4 * y - 20 * t * t - 16 * t + 2; };
    cout << "y(1) = " << taylor2(f, df, 0.0, 1.0, 0.1, 10) << "\n";
}

File: src/chapters/differential-equations/content/code/taylor2.go
package main

import "fmt"

// Second-order Taylor method. df is the total derivative f'(t,y) = f_t + f_y f.
func taylor2(f, df func(float64, float64) float64, t0, y0, h float64, n int) (float64, float64) {
	t, y := t0, y0
	for i := 0; i < n; i++ {
		y += h*f(t, y) + h*h/2*df(t, y)
		t += h
	}
	return t, y
}

func main() {
	// y' = 2y - 10t^2 + 2t  ->  f' = 4y - 20t^2 - 16t + 2
	f := func(t, y float64) float64 { return 2*y - 10*t*t + 2*t }
	df := func(t, y float64) float64 { return 4*y - 20*t*t - 16*t + 2 }
	_, y := taylor2(f, df, 0.0, 1.0, 0.1, 10)
	fmt.Println("y(1) =", y)
}
